README.md
Classical
##################
A simple cross-platform functional provider of classical inheritance for Javascript.

# Including in your code
## node.js
```javascript
require('classical')
```

Classical methods (Class, Extend, Public, etc.) will be placed in the global scope and can be used anywhere in your code.

If you would prefer not to use Classical at the global level, you will need to set ```process.env.CLASSICAL_PROTECTGLOBALS```
to ```true```, and then require classical in each class:

```javascript
var Classical = require('classical')
```

You can then use the classical variables as ```Classical.Class```, ```Classical.Extend```, ```Classical.Public```, etc.

## browser (unminified, global scope)
In the browser, Classical requires [require.js](http://requirejs.org/).

You must configure your requirejs to know where Classical's methods are located:

```javascript
requirejs.config({
    paths : {
        Classical           : '/path/to/classical/index',
        Class               : '/path/to/classical/src/Class',
        Interface           : '/path/to/classical/src/Interface'
    },
    deps: ['Classical', 'Class', 'Interface']
});
```

You can then use Classical as a normal requirement:

```javascript
define(function() {
    return Class(function() {});
});
```

## browser (unminified, local scope)
If you would prefer Classical not put itself at the window level, you will need to set ```window.CLASSICAL_PROTECTGLOBALS```
to ```true``` before requiring Classical.

You must still configure Classical the same way; however, there are some differences when using Classical later:

```javascript
define(['Classical'], function(Classical) {
    return Classical.Class(function() {});
});
```

## browser (minified, global scope)
**Classical does not currently work in minified format**

If you would like to use the minifed version of Classical, you must use a different requirejs config:

```javascript
requirejs.config({
    paths : {
        Classical           : '/path/to/classical/index.min.js'
    },
    deps: ['Classical']
}
```

## browser (minified, local scope)
**Classical does not currently work in minified format**

You can also use the minified version of Classical without affecting the window.  This is done the same way as before,
using the configuration for the minfied version.

# Examples
## Creating a Class
```javascript
require('classical');

var Mammal = Class(function() {
    this.name                           = Protected(null);

    this.constructor = Public(function(name) {
        this.name                       = name;
    });

    this.speak = Public(function(text) {
        console.log('%s says, "%s"', this.name, text);
    });
});
```

## Extension
```javascript
var Dog = Extend(Mammal, function() {
    this.breed                          = Protected(null);

    this.constructor = Public(function(name, breed) {
        this.breed                      = breed;
    });

    this.speak = Public(function() {
        this._super.speak('Woof!');
    });
});
```

## Instantiation
```javascript
var Spot = new Dog('Spot', 'Dalmation');
Spot.speak(); // Outputs: 'Spot says, "Woof!"' to the console.
```

# API
## Public
Defines a public member for a class.

### Parameters
```member```        The definition of the member.  This member will be publically available 
                    for all instances of the class and its children.

### Returns
```PublicMember```  A public member 

### Example
```javascript
this.foo = Public(function() {
    return 'foo';
});
```

## Protected
Defines a protected member for a class.

### Parameters
```member```        The definition of the member.  This member will be privately available for
                    all instances of the class and its children.

### Returns
```ProtectedMember```   A protected member

### Example
```javascript
this.foo = Protected(function() {
    return 'foo';
});
```

## Private
Defines a private member for a class.

### Parameters
```member```        The definition of the member.  This member will be privately available for
                    all instances of the class, but will be unavailable to its children.

### Returns
```PrivateMember``` A private member

### Example
```javascript
this.foo = Private(function() {
    return 'foo';
});
```

## Class
Defines a new Class.

### Parameters
```fn```            The definition of the class, represented as a function.

### Returns
```Class```         A factory to create instances of the class (using the new keyword)

### Example
```javascript
var Foo = Class(function() {
    this.foo = Protected(function() {
        return 'foo';
    });
    
    this.bar = Public(function() {
        return this.foo();
    });
});
var n   = new Foo;
console.log(n.foo()); // => 'foo'
```
## Extend
Extends an existing Classical or non-Classical class.

### Parameters
```ancestor```      The non-Classical class to inherit from.
```fn```            The definition of the class, represented as a function.

### Example
```javascript
var Baz = Inherit(require('event').EventEmitter, function() {
    this.constructor = Public(function() {
        this.on('baz', this.qux);
    });

    this.baz = Public(function() {
        this.emit('baz');
    });

    this.qux = Private(function() {
        console.log('foo');
    });
});
var n   = new Baz;
n.baz(); // => 'foo'
```

## Interface
Creates a new interface to be implemented.

### Parameters
```fn```            The definition of the interface, represented as a function.

### Example
```javascript
var Squee = Interface(function() {
    this.moog = Public(function(ap) {}); // A public method with one argument.
    this.dar = Protected(Class.BOOLEAN); // A protected boolean
    this.han = Public(Class.INT); // A public integer
    this.mik = Private(Class.STRING); // A private string
});
```

## Implement
Implements interfaces as a classical class.

### Parameters
```interfaces```    An interface (or several interfaces as an array).
```fn```            The definition of the implementing class, represented as a function.

### Example
```javascript
var Nop = Implement(Squee, function() {
    this.moog = Public(function(ap) {
        return Util.format(this.mik, ap, this.han, ap + this.han);
    });

    this.dar = Protected(true);
    this.han = Public(4);
    this.mik = Private("%s + %s = %s");
});
```


Changes.md
 * Changelog:
 *   - 2013-01-16: Added a means for the _internal object to access the public interface.
 *   - 2013-01-10: Fixed browser-based order of operations issues.
 *   - 2013-01-04: Cleaned up numerous bugs and created unit tests to watch for regression.
 *   - 2012-12-24: Added support for WebWorkers.
 *   - 2012-11-12: Members not given a visibility are now defaulted to public.
 *   - 2012-10-31: Added classical to bower.
 *   - 2012-10-31: Added a minified version.
 *   - 2012-10-31: Added instructions for browser usage.
 *   - 2012-10-26: Allowed extension of object ancestors.
 *   - 2012-10-23: Adapted frontend to include require.js hooks (uncomment to activate).
 *   - 2012-10-19: Minor bug in init script was causing conflicts with require statements in nodejs.
 *   - 2012-10-17: Rearchitected how Classical affects Globals, and added an option to protect the global namespace.
 *   - 2012-10-17: Moved TYPES to Interface, which actually uses them.
 *   - 2012-10-02: Added Inherit
 *   - 2012-09-28: Fixes for dereference method and handling of browser vs nodejs.
 *   - 2012-09-27: Complete rewrite to accomplish Public, Private, and Protected instances,
                   and to better handle method bindings. The need for method.call should be
                   almost nonexistent, now.
 *   - 2012-08-28: Converted to NPM registry.
 *   - 2012-08-23: SuperClass assignment was still going to parent instead of super.
 *   - 2012-08-21: Added interfaces and a method of implementing them.
 *   - 2012-08-21: Renamed .parent to .super; .parent is now deprecated.
 *   - 2012-08-21: When constructing an instance of a class, we now call the class's
                   parent constructors before calling the class constructor.


test/Interface.js
var Assert                              = require('assert');
var Classical                           = require('../');

suite('Interface');

var IFace;
var EIFace;
var Impl;

test('Create Interface', function() {
    IFace = Interface(function() {
        this.foo = Public(function(bar) {});
        this.bar = Protected(Classical.INT);
    });

    Assert.ok(IFace.name == 'Interface');
    try {
        var n = new IFace;
    }
    catch(e) {
        Assert.ok(e.message == 'Cannot instantiate interface.');
    }
});

test('Extend Interface', function() {
    EIFace = Extend(IFace, function() {
        this.bar = Protected(Classical.FLOAT);
        this.baz = Static(Public(Classical.INT));
    });

    Assert.ok(EIFace.name == 'Interface');
});

test('Implement Interface', function() {
    var badimplran = false;
    try {
        var BadImpl = Implement(EIFace, function() {
            this.foo = Public(function(bar) {});
            this.bar = Protected("1234");
        });
    }
    catch (e) {
        Assert.ok(e instanceof TypeError);
        badimplran = true;
    }
    finally {
        if (!badimplran) {
            Assert.ok(false, 'Did not throw TypeError as expected.');
        }
    }

    badimplran = false;
    try {
        BadImpl = Implement(EIFace, function() {
            this.foo = Public(function() {});
            this.bar = Protected(1234);
        });
    }
    catch (e) {
        Assert.ok(e instanceof TypeError);
        badimplran = true;
    }
    finally {
        if (!badimplran) {
            Assert.ok(false, 'Did not throw TypeError as expected.');
        }
    }

    // TODO: Visibility tests
    // TODO: Static tests

    Impl = Implement(EIFace, function() {
        this.foo = Public(function(bar) {});
        this.bar = Protected(1234);
    });

    Assert.ok(Impl.name == 'Class');
});

test/ClassTests.js
var Assert          = require('assert');
var Classical       = require('../index');

suite('Class');

var A;
var B;
var C;

afterEach(function() {
    process.removeAllListeners();
});

test('Initializing a class', function(done) {
    A                                   = require('./fixtures/A');

    process.on('A.constructor', function(val) {
        Assert.ok(val);
        done();
    });

    var a                               = new A;

    Assert.ok(a instanceof A);
});

test('Methods', function(done) {
    // Setup
    var tests = {
        constructor                     : false,
        one                             : false,
        two                             : false,
        three                           : false,
        four                            : false,
        fourAgain                       : false,
    };

    process.on('ClassTest', function(called) {
        if (called === 'four' && tests['four'] === true) {
            called                      = 'fourAgain';
        }

        tests[called]                   = true;

        try {
            Assert.deepEqual(tests, {
                constructor             : true,
                one                     : true,
                two                     : true,
                three                   : true,
                four                    : true,
                fourAgain               : true,
            });
            done();
        }
        catch (e) {}
    });

    process.on('A.constructor', function(val) {
        Assert.ok(val);
        process.emit('ClassTest', 'constructor');
    });

    process.on('A.one', function(val) {
        Assert.ok(val);
        process.emit('ClassTest', 'one');
    });

    process.on('A.two', function(val) {
        Assert.ok(val);
        process.emit('ClassTest', 'two');
    });

    process.on('A.three', function(val) {
        Assert.ok(val);
        process.emit('ClassTest', 'three');
    });

    process.on('A.four', function(val) {
        Assert.ok(val);
        process.emit('ClassTest', 'four');
    });

    // Execute
    var a                               = new A;

    Assert.equal(typeof a.one, 'function');
    Assert.notEqual(typeof a.two, 'function');
    Assert.notEqual(typeof a.three, 'function');
    Assert.equal(typeof A.four, 'function');
    Assert.equal(typeof a.four, 'function');

    a.one();
    a.four();
    A.four();
});

test('Extending Classes', function(done) {
    // Setup
    B                                   = require('./fixtures/B');

    var tests = {
        a : {
            constructor                 : false,
            one                         : false,
            two                         : false,
            three                       : false,
            four                        : false,
        },
        b : {
             constructor                : false,
             three                      : false,
             five                       : false,
        }
    };

    process.on('ClassTest', function(cl, called) {
        tests[cl][called]               = true;

        try {
            Assert.deepEqual(tests, {
                a : {
                    constructor                 : true,
                    one                         : true,
                    two                         : true,
                    three                       : false,
                    four                        : true,
                },
                b : {
                     constructor                : true,
                     three                      : true,
                     five                       : false,
                }
            });
            done();
        }
        catch (e) {}
    });

    process.on('A.constructor', function(val) {
        Assert.ok(val);
        process.emit('ClassTest', 'a', 'constructor');
    });

    process.on('A.one', function(val) {
        Assert.ok(val);
        process.emit('ClassTest', 'a', 'one');
    });

    process.on('A.two', function(val) {
        Assert.ok(val);
        process.emit('ClassTest', 'a', 'two');
    });

    process.on('A.three', function(val) {
        Assert.ok(val);
        process.emit('ClassTest', 'a', 'three');
    });

    process.on('A.four', function(val) {
        Assert.ok(val);
        process.emit('ClassTest', 'a', 'four');
    });

    process.on('B.constructor', function(val) {
        Assert.ok(val);
        process.emit('ClassTest', 'b', 'constructor');
    });

    process.on('B.three', function(val) {
        Assert.ok(val);
        process.emit('ClassTest', 'b', 'three');
    });

    process.on('B.five', function(val) {
        Assert.ok(val);
        process.emit('ClassTest', 'b', 'five');
    });

    // Execute
    var b                               = new B;

    Assert.equal(typeof b.one, 'function');
    Assert.notEqual(typeof b.two, 'function');
    Assert.notEqual(typeof b.three, 'function');
    Assert.equal(typeof B.four, 'function');
    Assert.equal(typeof b.four, 'function');
    Assert.equal(typeof b.five, 'function');

    b.one();
    b.four();
    // b.five();

    Assert.ok(b instanceof B);
    Assert.ok(b instanceof A);
});

test('Extending Objects', function(done) {
    // Setup
    C                                   = require('./fixtures/C');

    var tests = {
        one                             : false,
        two                             : false,
        C                               : false
    }
    
    process.on('ClassTest', function(called) {
        tests[called]                   = true;

        try {
            Assert.deepEqual(tests, {
                one                     : true,
                two                     : true,
                C                       : true
            });
            done();
        }
        catch(e) {};
    });

    process.on('test.one', function(val) { 
        Assert.ok(val);
        process.emit('ClassTest', 'one');
    });


    process.on('test.two', function(val) { 
        Assert.ok(val);
        process.emit('ClassTest', 'two');
    });

    process.on('C.one', function(val) {
        Assert.ok(val);
        process.emit('ClassTest', 'C');
    });

    // Execute
    var c                               = new C;
    c.one();
    c.two();
});

test('Accessing the public API', function() {
    var a                               = new A;
    Assert.deepEqual(a, a.getPublicApi());
});

test/fixtures/A.js
var A = Class(function() {
	this.constructor = Public(function() {
		process.emit('A.constructor', true);
	});

	this.one = Public(function() {
		process.emit('A.one', true);
		this.two();
		this.three();
	});

	this.two = Protected(function() {
		process.emit('A.two', true);
	});

	this.three = Private(function() {
		process.emit('A.three', true);
	});

	this.four = Static(Public(function() {
		process.emit('A.four', true);
	}));

    this.getPublicApi = Public(function() {
        return this._classical_public;
    });
});

module.exports = A;

index.js
/**
 * __          __   _ _   _                  _
 * \ \        / /  (_) | | |                | |
 *  \ \  /\  / / __ _| |_| |__    _ __   ___| |_
 *   \ \/  \/ / '__| | __| '_ \  | '_ \ / _ \ __|
 *    \  /\  /| |  | | |_| | | |_| | | |  __/ |_
 *     \/  \/ |_|  |_|\__|_| |_(_)_| |_|\___|\__|
 *
 * @created     2012-02-08
 * @edited      2012-01-04
 * @package     Libraries
 * @see         https://github.com/Writh/classical
 *
 * Copyright (C) 2012-2013 Kevin Kragenbrink <kevin@writh.net>
 *
 * Permission is hereby granted, free of charge, to any person obtaining a
 * copy of this software and associated documentation files (the "Software"),
 * to deal in the Software without restriction, including without limitation
 * the rights to use, copy, modify, merge, publish, distribute, sublicense,
 * and/or sell copies of the Software, and to permit persons to whom the
 * Software is furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
 * FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
 * DEALINGS IN THE SOFTWARE.I
 */
var a = function(global, require) {
    var version                             = '2.3.0';

    // Prevents shenanigans like loading classical twice.
    if (typeof process != 'undefined' && typeof process.versions != 'undefined') {
        if (typeof process.versions.classical != 'undefined') {
            if (version !== process.versions.classical) {
                throw new Error('Attempted to load classical ' + version + ', but version ' + process.versions.classical + ' is already loaded.');
            }
        }
        else {
            process.versions.classical      = version;
        }
    }

    if (typeof window == 'undefined') {
        var Class                           = require('./src/Class');
        var Interface                       = require('./src/Interface');
    }

    return global.Classical;
};

if (typeof window != 'undefined') {
    // requirejs
    define(['require'], a.bind(this, window));
}
else if (typeof module != 'undefined' && typeof module.exports != 'undefined') {
    // nodejs
    module.exports = a(global, require);
}
else {
    // web workers
    a(this, importScripts);
}


src/Class.js
/**
 * __          __   _ _   _                  _
 * \ \        / /  (_) | | |                | |
 *  \ \  /\  / / __ _| |_| |__    _ __   ___| |_
 *   \ \/  \/ / '__| | __| '_ \  | '_ \ / _ \ __|
 *    \  /\  /| |  | | |_| | | |_| | | |  __/ |_
 *     \/  \/ |_|  |_|\__|_| |_(_)_| |_|\___|\__|
 *
 * @created     2012-02-08
 * @edited      2013-01-16
 * @package     Libraries
 * @see         https://github.com/Writh/classical
 *
 * Copyright (C) 2012-2013 Kevin Kragenbrink <kevin@writh.net>
 *
 * Permission is hereby granted, free of charge, to any person obtaining a
 * copy of this software and associated documentation files (the "Software"),
 * to deal in the Software without restriction, including without limitation
 * the rights to use, copy, modify, merge, publish, distribute, sublicense,
 * and/or sell copies of the Software, and to permit persons to whom the
 * Software is furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
 * FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
 * DEALINGS IN THE SOFTWARE.I
 */
var b = function(global, require, module) {
    if (typeof module.exports == 'undefined') { module.exports = {}; }

    /**
     * Defines a new Class.
     *
     * @global
     * @constructor
     */
    module.exports.Class = function(fn) { return defineClass(fn); };

    /**
     * Defines a new Class using a non-Classical class as the baseclass.
     *
     * @global
     * @constructor
     */
    module.exports.Extend = function(ancestor, fn) { return defineClass(fn, undefined, ancestor); };

    /**
     * Creates a public member of a Class.
     *
     * @global
     * @param   member
     * @return  {Object}
     * @constructor
     */
    module.exports.Public = function Public(member) { return setVisibility(member); };

    /**
     * Creates a private member of a Class.
     *
     * @global
     * @param   member
     * @return  {Object}
     * @constructor
     */
    module.exports.Private = function Private(member) { return setVisibility(member); };

    /**
     * Creates a protected member of a Class.
     *
     * @global
     * @param   member
     * @return  {Object}
     * @constructor
     */
    module.exports.Protected = function Protected(member) { return setVisibility(member); };

    /**
     * Creates a static member of a Class.
     *
     * @global
     * @param   member
     * @return  {Object}
     * @constructor
     */
    module.exports.Static = function Static(member) { return setStatic(member); };

    /**
     * The base class prototype.
     * @constructor
     */
    module.exports.BaseClass = function BaseClass() {};

    /***************************************
     * INTERNALS
     *
     * The following code is not a part of the public API for Classical, and
     * should not be altered or exposed to external scripts.
     ***************************************/
    var BaseClass                       = module.exports.BaseClass;

    /**
     * Implements Classical inheritance by wrapping prototypal inheritance.
     *
     * @param   {Function}      fn
     * @param   {Object}        _super
     * @param   {Function}      ancestor
     * @return  {ClassFactory}
     * @constructor
     * @global
     */
    var defineClass = function(fn, _super, ancestor) {
        var member;

        // Get a _preInstance.
        var _preInstance                = new fn;

        // Extend the _preInstance with members from the inheritted class, if any.
        if (typeof ancestor == 'function'
            ||  (typeof ancestor == 'object' && Object.prototype.toString.call(ancestor) == '[object Object]')) {

            if (typeof ancestor._classical_extend == 'function') {
                return ancestor._classical_extend(fn);
            }
            else {
                var AncestralClass      = (typeof ancestor == 'function') ? new ancestor : dereference(ancestor);

                for (member in AncestralClass) {
                    // WARNING: Not using hasOwnProperty here because Node EventEmitter does not have its
                    //          methods as its own properties. This creates a pretty big opening if
                    //          someone foolishly modifies the Object prototype.
                    AncestralClass[member]                  = Public(AncestralClass[member]);
                    if (typeof _preInstance[member] == 'undefined') {
                        _preInstance[member]                = AncestralClass[member];
                    }
                }
            }
        }

        // Create a copy of the SuperClass to set the prototype.
        initializing                    = true;
        var SuperClass                  = (typeof this == 'function') ? this : BaseClass;
        if (typeof ancestor == 'function') {
            SuperClass.prototype        = AncestralClass;
        }
        var prototype                   = new SuperClass;
        initializing                    = false;

        // Establish a baseline prototype chain.
        var base                        = function() {};
        base.prototype                  = prototype;

        // Extend the _preInstance with the members from the SuperClass
        if (_super !== undefined) {
            for (member in _super) {
                if (member != 'constructor' && _super.hasOwnProperty(member)) {
                    if ((_super[member]._visibility == 'Public' || _super[member]._visibility == 'Protected')
                        && typeof _preInstance[member] == 'undefined') {

                        _preInstance[member]                = _super[member];
                    }
                }
            }

            // Add the _super to the _preInstance, for construction and this._super purposes.
            _preInstance._super         = _super;
        }
        else if (AncestralClass !== undefined) {
            _preInstance._super         = AncestralClass;
        }

        return getClassFactory(base, prototype, _preInstance, arguments.callee);
    };

    var getClassFactory = function(base, prototype, _preInstance, extend) {
        var member;

        // Set up the ClassFactory.  This is what actually gets instantiated.
        var ClassFactory = function Class() {
            var _instance               = new base;
            var _public                 = new base;

            _instance._classical_public= _public;

            if (!initializing) {
                assemble(_instance, _instance, _public, _preInstance);
                construct(_instance, _instance, arguments);
            }

            return _public;
        };

        // Bind the ClassFactory to the prototype (for the inheritance chain).
        ClassFactory.prototype          = prototype;

        // Set up static methods.
        for (member in _preInstance) {
            if (member != '_super' && _preInstance.hasOwnProperty(member)) {
                // While we're at it, let's also make sure visibility has been set.
                if (!_preInstance[member].hasOwnProperty('_visibility')) {
                    _preInstance[member]                    = Public(_preInstance[member]);
                }

                if (_preInstance[member]._static === true && _preInstance[member]._visibility == 'Public') {
                    ClassFactory[member]                    = copyValue(_preInstance[member]._value, ClassFactory, _preInstance[member]._static);
                }
            }
        }

        // Setup a method to extend the base class.
        ClassFactory._classical_extend = function(newfn) {
            return extend.call(ClassFactory, newfn, _preInstance);
        };

        // Assembles the instance and its public accessors.
        var assemble = function(_context, _instance, _public, _preInstance) {
            var member;

            if (typeof _preInstance != 'undefined' && typeof _preInstance._super != 'undefined') {
                _instance._super                            = new base;
                assemble(_context, _instance._super, {}, _preInstance._super);
            }

            for (member in _preInstance) {
                if (member != '_super' && _preInstance.hasOwnProperty(member)) {
                    if (_preInstance[member]._visibility != 'Private' || _context === _instance) {
                        _instance[member]                   = copyValue(_preInstance[member]._value, _context,
                                                                        _preInstance[member]._static);

                        if (_preInstance[member]._visibility == 'Public') {
                            _public[member]                 = _instance[member];
                        }
                    }
                }
            }
        };

        // Recursively calls the constructors for the class, starting with the ancestor
        // and working towards the instance.
        var construct = function(context, target, arguments) {
            // Call the SuperClass constructors first.
            if (target._super instanceof BaseClass) {
                construct(context, target._super, arguments);
            }

            // Call the instance constructor.
            if (typeof target.constructor == 'function' && target.hasOwnProperty('constructor')) {
                target.constructor.apply(context, arguments);
            }
        };

        return ClassFactory;
    };

    /**
     * Performs a deep clone to dereference an object.
     *
     * @param   {*}     source
     * @return  {*}
     */
    var dereference = function(source) {
        var target;

        if (typeof source == 'object' && source !== null) {
            target = (typeof source.length == 'number') ? [] : {};

            for (var i in source) {
                if (source.hasOwnProperty(i)) {
                    target[i]           = dereference(source[i]);
                }
            }
        }
        else {
            target                      = source;
        }

        return target;
    };

    /**
     * Used to prevent calling the constructor during ClassFactory creation.
     * @type    {Boolean}
     */
    var initializing = false;

    /**
     * Returns a dereferenced copy of a value, bound to the target if it is a function.
     *
     * @param   {*}         member
     * @param   {Object}    target
     * @param   {Boolean}   _static
     * @return  {*}
     */
    var copyValue = function(member, target, _static) {
        if (typeof member == 'function') {
            return member.bind(target);
        }
        else {
            return (!_static ? dereference(member) : member);
        }
    };

    /**
     * A common interface for the Public, Private, and Protected APIs.
     * @param   {*}         member
     * @return  {Object}
     */
    var setVisibility = function(member) {
        return {
            _value                      : member,
            _visibility                 : arguments.callee.caller.name,
            _static                     : false
        }
    };

    /**
     * A wrapper around the Public, Private, and Protected APIs to make a method static.
     * @param   {Object}    member
     * @return  {Object}
     */
    var setStatic = function(member) {
        member._static                  = true;
        return member;
    };

    // Set up the global Classical variables.
    global.Classical                    = global.Classical || {};
    global.Classical.Class              = module.exports.Class;
    global.Classical.Extend             = module.exports.Extend;
    global.Classical.Public             = module.exports.Public;
    global.Classical.Private            = module.exports.Private;
    global.Classical.Protected          = module.exports.Protected;
    global.Classical.Static             = module.exports.Static;

    // Register classical globals.
    if ((typeof process != 'undefined' && typeof process.env != 'undefined' && process.env.CLASSICAL_PROTECTGLOBALS !== true)
        ||  (typeof window != 'undefined' && window.CLASSICAL_PROTECTGLOBALS !== true)
        ||  (typeof global != 'undefined' && global.CLASSICAL_PROTECTGLOBALS !== true)) {
        global.Class                    = global.Classical.Class;
        global.Extend                   = global.Classical.Extend;
        global.Public                   = global.Classical.Public;
        global.Private                  = global.Classical.Private;
        global.Protected                = global.Classical.Protected;
        global.Static                   = global.Classical.Static;
    }

    return module.exports;
};


if (typeof window != 'undefined') {
    // requirejs
    define(['require', 'module', 'Classical'], b.bind(this, window));
}
else if (typeof module != 'undefined' && typeof module.exports != 'undefined') {
    // nodejs
    b(global, require, module);
}
else {
    // web workers
    b(this, null, this);
}
